CONTRIBUTING.md

```diff
@@ -1,44 +1,154 @@
+Follow this guide to ensure your effort is put to the best possible use.
+
+Contributing to open source projects requires a shift in mindset.
+Often contributors have had most of their experience in corporate software development.
+The dynamics of open source software projects differ in important ways from the corporate domain.
+The primary difference is that **we are all volunteers**.
+There is very little obligation for anybody to perform a specific piece of work.
+In particular, if somebody produces some changes that are merged and then found to cause a bug, there is no obligation for the submitter to fix that bug like there is in a corporate setting.
+The submitter is well within their rights to not volunteer any more of their labor.
+This means the threshold for accepting changes to open source projects is usually higher than it is in corporate settings, because no feedback loop exists to ensure the contributor remains responsible for their changes going forward.
+Once a change is merged it becomes the responsibility of the maintainers, who are also volunteers and so wish to minimize their maintenance burden.
+Above all, please remember to be kind to everybody!
+You can review our Code of Conduct [here](.github/CODE_OF_CONDUCT.md).
+
+The Contribution Process & Style
+--------------------------------
+
+The most important step is to engage with existing developers & maintainers before beginning work.
+The best way to do this is to comment on an issue you want to fix, or open a new issue if a suitable candidate does not exist.
+It is also very helpful to join the [monthly online community meeting](https://groups.google.com/g/tlaplus/c/CpAEnrf-DHQ/m/YrORpIfSBwAJ) to introduce yourself and speak with people directly.
+The development team will help you better understand the scope & difficulty of the proposed changes, along with what parts of the codebase they'll touch and what additional tests are required to safely merge them.
+Building & maintaining consensus around these aspects will ensure a smooth review & merge process.
+
+To begin work on your change, fork this repository then create an appropriately-named branch to track your work.
+You will eventually submit a PR between your feature branch and the master branch of this repository, at which point developers & maintainers will review your changes.
+After integrating this feedback your PR will then hopefully be merged.
+
+When creating and submitting your changes, closely follow these guidelines:
+ * **Be surgical:** What is the smallest diff you can make that still accomplishes your goal?
+ Avoid the urge to fix unrelated stylistic issues in a file you are changing.
+ * **Write tests first:** If you're changing a specific part of the codebase, first ensure it has good test coverage; if it does, write a few more tests yourself anyway!
+ Writing tests is a great way to learn how the code functions.
+ Submit your tests in a separate PR so their benefits can be realized immediately.
```

Comment:
I would add "If possible" to the start of this sentence; if the change is a bug fix, it doesn't make much sense to merge a failing test case first (even if it might make sense to _write_ the failing test first).